### PR TITLE
Refactor terminology lib

### PR DESCRIPTION
**Pull request by nirmal070125**
- improve the code by avoiding repeated code
- make terminology lib extensible by providing an interface to implement a custom one. as part of issue https://github.com/wso2-enterprise/open-healthcare/issues/1564
- following best practices of Ballerina
- optimized the code avoiding unnecessary locks and concurrency controls
- handle cases like a Valueset contains a list of valuesets, codesystems etc.
- add more tests to bring the code coverage to 91.46%.

### CODE REVIEW

File: terminology/inmemory_terminology.bal [hunk 1]
```diff
@@ -0,0 +1,439 @@
+import ballerina/http;
```
Comment (sameeragunarathne) [inline terminology/inmemory_terminology.bal:1]: Missing licence

File: terminology/inmemory_terminology.bal [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+                    resultedValueSetMap[key] = v;
+                }
+            }
+        }
+
+        // Return the updated codeSystemMap
+        return resultedValueSetMap;
+    }
+
+    public isolated function addCodeSystem(r4:CodeSystem codeSystem) returns r4:FHIRError? {
```
Comment (sameeragunarathne) [inline terminology/inmemory_terminology.bal:107]: Let's add bal docs for public functions

Comment (nirmal070125) [reply terminology/inmemory_terminology.bal:107]: This is inherited from the Terminology interface.

File: terminology/inmemory_terminology.bal (hunk 1 above)
Comment (nirmal070125) [reply terminology/inmemory_terminology.bal:1]: ack.

File: terminology/terminology.bal [hunk 3]
```diff
@@ -107,24 +198,220 @@ public isolated function codeSystemLookUp(r4:code|r4:Coding|r4:CodeableConcept c
...
+            return r4:createFHIRError(
+                        string `Cannot find a ValueSet for system: ${system.toString()}`,
+                        r4:ERROR,
+                        r4:INVALID_REQUIRED,
+                        errorType = r4:PROCESSING_ERROR,
+                        httpStatusCode = http:STATUS_BAD_REQUEST
+                    );
+        }
+    }
+    else {
```
Comment (sameeragunarathne) [inline terminology/terminology.bal:345]: Formatting issue

File: terminology/utils.bal [hunk 4]
```diff
@@ -0,0 +1,536 @@
+import ballerina/http;
```
Comment (sameeragunarathne) [inline terminology/utils.bal:1]: Missing licence

File: terminology/terminology.bal (hunk 3 above)
Comment (nirmal070125) [reply terminology/terminology.bal:345]: sorry - not clear

Comment (nirmal070125) [reply terminology/terminology.bal:345]: formatted using plugin.

**Review by sameeragunarathne — APPROVED** [review]

**Pull request merged**